### DISCUSSION TITLE
Get 403 Forbidden when proxy by nginx

### DISCUSSION THREAD

**Discussion by lingen** [open-webui/open-webui#291] [General]
I just want to proxy ollama-webui by nginx. 
And there is the config 
```nginx

server
{
	listen 3389;
	server_name ai.myai.com;

	location /
	{

		proxy_redirect off;
		proxy_pass http://127.0.0.1:8082;
		proxy_http_version 1.1;
		proxy_set_header Upgrade $http_upgrade;
		proxy_set_header Connection "upgrade";
		proxy_set_header Host $host;
		proxy_set_header X-Real-IP $remote_addr;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_next_upstream off;


	}
}

```

Also, I started the ollama-webui in docker.
```shell
docker run -d -p 8082:8080 --add-host=host.docker.internal:host-gateway --name ollama-webui --restart always ghcr.io/ollama-webui/ollama-webui:main
```

In fact, When I started ollama-webui in the 3389 port, everything was ok. But when I proxy it by nginx, there is a 403 Forbidden error. 

Do anyone know what the problems are?

<img width="644" alt="image" src="https://github.com/ollama-webui/ollama-webui/assets/2062865/ca43686d-86e9-4742-8b1f-baae2fa9d9ed">

**Comment by oliverbob** [comment]
Have you tried doing it with this [Apache proxy example](https://github.com/ollama-webui/ollama-webui/issues/256)?

**Comment by lingen** [comment]
Thank you very much. I will try it.

**Comment by lingen** [comment]
@oliverbob 

OK, I solve this problem. 

The blew config of the Nginx proxy is ok. 
```nginx
server {
    listen 3389;
    server_name ai.myai.com;

    location / {
      proxy_pass http://127.0.0.1:8082;

    }

    location /ollama/api {
      proxy_buffering off;
      proxy_set_header Origin '';
      proxy_set_header Referer '';
      proxy_pass http://127.0.0.1:11434/api;

    }

    location /api/v1 {
     proxy_pass http://127.0.0.1:8082/api/v1;
    }
}
```

And there two lines are very important

```shell
      proxy_set_header Origin '';
      proxy_set_header Referer '';
```

But I have no idea why this happened. And I thought this was not Ollama-webui's problem. Maybe it's ollama server's problem.

Because the 403 error is generated by Ollama Server but not Ollama-webui.

Thanks. I appreciate your help.

**Comment by oliverbob** [comment]
@lingen, it is not actually an Ollama problem, but it is a Mixed content cors policy issue.

The solution to this, as I outlined was to proxy:

location /ollama/api to point to ollama
location /api/v1 to point to the UI.

I suggest someone should at least make a PR about changes/addition to the docs.

I could probably make a docs on this for apache, and you could perhaps make a doc nginx.

Also, there should be a doc to accommodate nGrok users for people who are comfortable with a third party solution.

I should also point out that at least in the new commit of the dev branch about RBAC, there is an auths error/bug at https://ollama-remote.tld/api/v1/auths when trying to reload the page, requiring the user to re-authenticate.  But it is not a problem though when running via localhost, but defeats the purpose of the multiuser fullstack software support goal.

**Comment by rmkni** [comment]
may be related https://github.com/open-webui/open-webui/discussions/2379

**Comment by PyroGenesis** [comment]
For me, this configuration fixed the `403` error:

```nginx
http {

    upstream ollama {
        server               localhost:11434;
    }

    server {
        listen              8000 ssl;
        server_name         ollama.mydomain.com;

        # these options are only needed for ssl
        ssl_certificate     "path/to/cert.crt";
        ssl_certificate_key "path/to/private.key";
        ssl_protocols       TLSv1 TLSv1.1 TLSv1.2;
        ssl_ciphers         HIGH:!aNULL:!MD5;

        location / {
            proxy_set_header  Host "localhost";
            proxy_set_header  X-Real-IP "127.0.0.1";
            proxy_set_header  X-Forwarded-For "127.0.0.1";
            proxy_set_header  X-Forwarded-Proto $scheme;
            proxy_pass        http://ollama;  # Forward request to the actual web service
        }
    }
```

The solution was to simulate the request coming from `localhost`, since I didn't want to set the Ollama origin as `*`.
